File: scripts/calibration/imu_calibration.py
#!/usr/bin/env python3
"""
IMU Sensor Calibration System

Calibrates accelerometers, gyroscopes, and magnetometers using:
- Static position calibration for accelerometers
- Zero-velocity calibration for gyroscopes
- Ellipsoid fitting for magnetometers

Similar to PT calibration pattern from external FSW.
"""

import numpy as np
import json
import csv
import time
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass, asdict
from pathlib import Path
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class SensorType(Enum):
    ACCELEROMETER = "accelerometer"
    GYROSCOPE = "gyroscope"
    MAGNETOMETER = "magnetometer"


@dataclass
class CalibrationPoint:
    """Single calibration data point"""

    raw_value: np.ndarray  # Raw sensor reading [3x1]
    reference_value: np.ndarray  # Reference/true value [3x1]
    temperature: float  # Temperature [°C]
    timestamp: float
    uncertainty: float = 0.01


@dataclass
class CalibrationParams:
    """Calibration parameters"""

    bias: np.ndarray  # Bias vector [3x1]
    scale_matrix: np.ndarray  # Scale factor matrix [3x3]
    misalignment: np.ndarray  # Misalignment matrix [3x3]
    temperature_coeff: float = 0.0
    reference_temperature: float = 25.0

    # Quality metrics
    bias_uncertainty: float = 0.0
    scale_uncertainty: float = 0.0
    calibration_quality: float = 0.0

    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON serialization"""
        return {
            "bias": self.bias.tolist(),
            "scale_matrix": self.scale_matrix.tolist(),
            "misalignment": self.misalignment.tolist(),
            "temperature_coeff": self.temperature_coeff,
            "reference_temperature": self.reference_temperature,
            "bias_uncertainty": self.bias_uncertainty,
            "scale_uncertainty": self.scale_uncertainty,
            "calibration_quality": self.calibration_quality,
        }

    @classmethod
    def from_dict(cls, data: Dict) -> "CalibrationParams":
        """Create from dictionary"""
        return cls(
            bias=np.array(data["bias"]),
            scale_matrix=np.array(data["scale_matrix"]),
            misalignment=np.array(data["misalignment"]),
            temperature_coeff=data.get("temperature_coeff", 0.0),
            reference_temperature=data.get("reference_temperature", 25.0),
            bias_uncertainty=data.get("bias_uncertainty", 0.0),
            scale_uncertainty=data.get("scale_uncertainty", 0.0),
            calibration_quality=data.get("calibration_quality", 0.0),
        )


class IMUCalibrator:
    """Base IMU sensor calibrator"""

    def __init__(self, sensor_type: SensorType, sensor_id: str):
        self.sensor_type = sensor_type
        self.sensor_id = sensor_id
        self.calibration_points: List[CalibrationPoint] = []
        self.params: Optional[CalibrationParams] = None

    def add_calibration_point(self, point: CalibrationPoint):
        """Add calibration data point"""
        self.calibration_points.append(point)

    def calibrate(self) -> CalibrationParams:
        """Perform calibration"""
        raise NotImplementedError

    def apply_calibration(
        self, raw_value: np.ndarray, temperature: float = 25.0
    ) -> np.ndarray:
        """Apply calibration to raw reading"""
        if self.params is None:
            raise ValueError("Calibration not performed")

        # Temperature compensation
        temp_scale = 1.0
        if abs(self.params.temperature_coeff) > 1e-6:
            temp_diff = temperature - self.params.reference_temperature
            temp_scale = 1.0 + self.params.temperature_coeff * temp_diff

        # Apply calibration: calibrated = scale_matrix * (raw - bias) * temp_scale
        corrected = raw_value - self.params.bias
        corrected *= temp_scale
        calibrated = self.params.scale_matrix @ corrected

        return calibrated

    def save_calibration(self, filepath: str):
        """Save calibration parameters to JSON file"""
        if self.params is None:
            raise ValueError("No calibration parameters to save")

        data = {
            "sensor_type": self.sensor_type.value,
            "sensor_id": self.sensor_id,
            "calibration_params": self.params.to_dict(),
            "num_points": len(self.calibration_points),
        }

        with open(filepath, "w") as f:
            json.dump(data, f, indent=2)

        logger.info(f"Saved calibration to {filepath}")

    def load_calibration(self, filepath: str):
        """Load calibration parameters from JSON file"""
        with open(filepath, "r") as f:
            data = json.load(f)

        self.params = CalibrationParams.from_dict(data["calibration_params"])
        logger.info(f"Loaded calibration from {filepath}")


class AccelerometerCalibrator(IMUCalibrator):
    """Accelerometer calibrator using static position method"""

    def __init__(self, sensor_id: str):
        super().__init__(SensorType.ACCELEROMETER, sensor_id)
        self.gravity = 9.81  # m/s²

    def calibrate(self) -> CalibrationParams:
        """Calibrate accelerometer using static positions"""
        if len(self.calibration_points) < 6:
            raise ValueError("Need at least 6 positions for accelerometer calibration")

        # Estimate bias as mean of all readings
        # (assuming gravity cancels out over all orientations)
        bias = np.mean([p.raw_value for p in self.calibration_points], axis=0)

        # Estimate scale matrix using least squares
        # For each point: reference = scale_matrix * (raw - bias)
        A = []
        b = []

        for point in self.calibration_points:
            corrected = point.raw_value - bias
            # Build linear system for scale matrix
            for i in range(3):
                row = np.zeros(9)
                for j in range(3):
                    row[i * 3 + j] = corrected[j]
                A.append(row)
                b.append(point.reference_value[i])

        A = np.array(A)
        b = np.array(b)

        # Solve least squares
        scale_vec = np.linalg.lstsq(A, b, rcond=None)[0]

        # Reshape to 3x3 matrix
        scale_matrix = scale_vec.reshape(3, 3)

        # Compute quality metrics
        residuals = []
        for point in self.calibration_points:
            calibrated = self.apply_calibration(point.raw_value, point.temperature)
            error = np.linalg.norm(calibrated - point.reference_value)
            residuals.append(error)

        avg_error = np.mean(residuals)
        quality = 1.0 / (1.0 + avg_error)

        self.params = CalibrationParams(
            bias=bias,
            scale_matrix=scale_matrix,
            misalignment=np.eye(3),  # Included in scale_matrix
            calibration_quality=quality,
            bias_uncertainty=np.std(
                [p.raw_value for p in self.calibration_points], axis=0
            ).mean(),
        )

        return self.params


class GyroscopeCalibrator(IMUCalibrator):
    """Gyroscope calibrator using zero-velocity method"""

    def __init__(self, sensor_id: str):
        super().__init__(SensorType.GYROSCOPE, sensor_id)

    def calibrate(self) -> CalibrationParams:
        """Calibrate gyroscope using zero-velocity periods"""
        if len(self.calibration_points) < 10:
            raise ValueError("Need at least 10 zero-velocity readings")

        # Estimate bias as mean (should be zero during stationary periods)
        bias = np.mean([p.raw_value for p in self.calibration_points], axis=0)

        # Estimate scale factors from variance
        # Higher variance indicates lower scale factor confidence
        variances = np.var([p.raw_value for p in self.calibration_points], axis=0)

        # Scale matrix: identity with scale factors inversely related to variance
        scale_matrix = np.eye(3)
        # For now, assume unity scale (would need rotation calibration for full scale)

        # Compute quality
        residuals = []
        for point in self.calibration_points:
            calibrated = self.apply_calibration(point.raw_value, point.temperature)
            error = np.linalg.norm(calibrated - point.reference_value)
            residuals.append(error)

        avg_error = np.mean(residuals)
        quality = 1.0 / (1.0 + avg_error)

        self.params = CalibrationParams(
            bias=bias,
            scale_matrix=scale_matrix,
            misalignment=np.eye(3),
            calibration_quality=quality,
            bias_uncertainty=np.std(
                [p.raw_value for p in self.calibration_points], axis=0
            ).mean(),
        )

        return self.params


class MagnetometerCalibrator(IMUCalibrator):
    """Magnetometer calibrator using ellipsoid fitting"""

    def __init__(self, sensor_id: str):
        super().__init__(SensorType.MAGNETOMETER, sensor_id)
        self.reference_field_magnitude = 1.0  # Normalized

    def calibrate(self) -> CalibrationParams:
        """Calibrate magnetometer using ellipsoid fitting"""
        if len(self.calibration_points) < 6:
            raise ValueError(
                "Need at least 6 orientations for magnetometer calibration"
            )

        # Hard iron correction: bias is center of ellipsoid
        # Estimate as mean of all readings
        bias = np.mean([p.raw_value for p in self.calibration_points], axis=0)

        # Soft iron correction: scale and misalignment
        # For now, use simplified approach
        # Full ellipsoid fitting would solve for 9 parameters

        # Compute scale factors from variance in each axis
        corrected_readings = [p.raw_value - bias for p in self.calibration_points]
        variances = np.var(corrected_readings, axis=0)

        # Normalize to reference field magnitude
        mean_magnitude = np.mean([np.linalg.norm(cr) for cr in corrected_readings])
        if mean_magnitude > 0:
            scale_factor = self.reference_field_magnitude / mean_magnitude
        else:
            scale_factor = 1.0

        scale_matrix = np.eye(3) * scale_factor

        # Compute quality
        residuals = []
        for point in self.calibration_points:
            calibrated = self.apply_calibration(point.raw_value, point.temperature)
            error = np.linalg.norm(calibrated - point.reference_value)
            residuals.append(error)

        avg_error = np.mean(residuals)
        quality = 1.0 / (1.0 + avg_error)

        self.params = CalibrationParams(
            bias=bias,
            scale_matrix=scale_matrix,
            misalignment=np.eye(3),
            calibration_quality=quality,
            bias_uncertainty=np.std(
                [p.raw_value for p in self.calibration_points], axis=0
            ).mean(),
        )

        return self.params


class IMUCalibrationSystem:
    """Complete IMU calibration system"""

    def __init__(self):
        self.accel_calibrator: Optional[AccelerometerCalibrator] = None
        self.gyro_calibrator: Optional[GyroscopeCalibrator] = None
        self.mag_calibrator: Optional[MagnetometerCalibrator] = None

    def create_calibrator(
        self, sensor_type: SensorType, sensor_id: str
    ) -> IMUCalibrator:
        """Create calibrator for sensor type"""
        if sensor_type == SensorType.ACCELEROMETER:
            self.accel_calibrator = AccelerometerCalibrator(sensor_id)
            return self.accel_calibrator
        elif sensor_type == SensorType.GYROSCOPE:
            self.gyro_calibrator = GyroscopeCalibrator(sensor_id)
            return self.gyro_calibrator
        elif sensor_type == SensorType.MAGNETOMETER:
            self.mag_calibrator = MagnetometerCalibrator(sensor_id)
            return self.mag_calibrator
        else:
            raise ValueError(f"Unknown sensor type: {sensor_type}")

    def calibrate_all(self) -> Dict[str, CalibrationParams]:
        """Calibrate all sensors"""
        results = {}

        if self.accel_calibrator and len(self.accel_calibrator.calibration_points) >= 6:
            results["accelerometer"] = self.accel_calibrator.calibrate()

        if self.gyro_calibrator and len(self.gyro_calibrator.calibration_points) >= 10:
            results["gyroscope"] = self.gyro_calibrator.calibrate()

        if self.mag_calibrator and len(self.mag_calibrator.calibration_points) >= 6:
            results["magnetometer"] = self.mag_calibrator.calibrate()

        return results

    def save_all_calibrations(self, base_path: str):
        """Save all calibrations"""
        if self.accel_calibrator and self.accel_calibrator.params:
            self.accel_calibrator.save_calibration(f"{base_path}_accel.json")

        if self.gyro_calibrator and self.gyro_calibrator.params:
            self.gyro_calibrator.save_calibration(f"{base_path}_gyro.json")

        if self.mag_calibrator and self.mag_calibrator.params:
            self.mag_calibrator.save_calibration(f"{base_path}_mag.json")


def main():
    """Example usage"""
    import argparse

    parser = argparse.ArgumentParser(description="IMU Calibration System")
    parser.add_argument(
        "--sensor-type", choices=["accel", "gyro", "mag"], required=True
    )
    parser.add_argument("--sensor-id", default="imu_0")
    parser.add_argument("--data-file", help="CSV file with calibration data")
    parser.add_argument("--output", help="Output calibration file")

    args = parser.parse_args()

    # Create calibrator
    calib_system = IMUCalibrationSystem()

    if args.sensor_type == "accel":
        calibrator = calib_system.create_calibrator(
            SensorType.ACCELEROMETER, args.sensor_id
        )
    elif args.sensor_type == "gyro":
        calibrator = calib_system.create_calibrator(
            SensorType.GYROSCOPE, args.sensor_id
        )
    else:
        calibrator = calib_system.create_calibrator(
            SensorType.MAGNETOMETER, args.sensor_id
        )

    # Load data from CSV or use interactive mode
    if args.data_file:
        # Load from CSV
        # Format: raw_x, raw_y, raw_z, ref_x, ref_y, ref_z, temperature
        with open(args.data_file, "r") as f:
            reader = csv.reader(f)
            for row in reader:
                raw = np.array([float(row[0]), float(row[1]), float(row[2])])
                ref = np.array([float(row[3]), float(row[4]), float(row[5])])
                temp = float(row[6]) if len(row) > 6 else 25.0

                point = CalibrationPoint(
                    raw_value=raw,
                    reference_value=ref,
                    temperature=temp,
                    timestamp=time.time(),
                )
                calibrator.add_calibration_point(point)
    else:
        print("Interactive calibration mode")
        print("Enter calibration points (or 'done' to finish):")
        # Interactive input would go here

    # Perform calibration
    try:
        params = calibrator.calibrate()
        print(f"Calibration successful!")
        print(f"  Bias: {params.bias}")
        print(f"  Quality: {params.calibration_quality:.3f}")

        if args.output:
            calibrator.save_calibration(args.output)
    except Exception as e:
        print(f"Calibration failed: {e}")


if __name__ == "__main__":
    main()

File: daq-server/diablo_server/lib/src/calibration/IMUCalibration.cpp
#include "../include/calibration/IMUCalibration.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>

namespace fsw {
namespace calibration {

IMUCalibration::IMUCalibration(SensorType type)
    : sensor_type_(type), status_(CalibrationStatus::NOT_CALIBRATED) {
    params_ = CalibrationParams{};
    params_.scale_matrix = Eigen::Matrix3d::Identity();
    params_.misalignment = Eigen::Matrix3d::Identity();
}

IMUCalibration::~IMUCalibration() = default;

void IMUCalibration::initialize(const CalibrationParams& initial_params) {
    std::lock_guard<std::mutex> lock(mutex_);
    params_ = initial_params;
    status_ = CalibrationStatus::NOT_CALIBRATED;
    calibration_data_.clear();
    recent_readings_.clear();
}

void IMUCalibration::addCalibrationPoint(const CalibrationPoint& point) {
    std::lock_guard<std::mutex> lock(mutex_);
    calibration_data_.push_back(point);
}

bool IMUCalibration::calibrate() {
    std::lock_guard<std::mutex> lock(mutex_);

    if (calibration_data_.empty()) {
        status_ = CalibrationStatus::FAILED;
        return false;
    }

    status_ = CalibrationStatus::CALIBRATING;

    // Extract raw readings
    std::vector<RawReading> readings;
    for (const auto& point : calibration_data_) {
        readings.push_back(point.raw);
    }

    // Perform calibration based on sensor type
    bool success = false;
    switch (sensor_type_) {
        case SensorType::ACCELEROMETER:
            success = calibrateAccelerometer(readings);
            break;
        case SensorType::GYROSCOPE:
            success = calibrateGyroscope(readings);
            break;
        case SensorType::MAGNETOMETER:
            // For magnetometer, need reference field
            if (calibration_data_.size() > 0) {
                Eigen::Vector3d ref_field = calibration_data_[0].reference;
                success = calibrateMagnetometer(readings, ref_field);
            }
            break;
    }

    if (success) {
        params_.calibration_quality = computeCalibrationQuality();
        status_ = CalibrationStatus::CALIBRATED;
    } else {
        status_ = CalibrationStatus::FAILED;
    }

    return success;
}

bool IMUCalibration::calibrateAccelerometer(const std::vector<RawReading>& readings) {
    if (readings.size() < 6) {
        std::cerr << "[IMUCalibration] Need at least 6 positions for accelerometer calibration"
                  << std::endl;
        return false;
    }

    // Estimate bias as mean of readings (assuming gravity cancels out)
    estimateBias(readings);

    // For accelerometer, use static calibration
    return calibrateAccelStatic(readings);
}

bool IMUCalibration::calibrateGyroscope(const std::vector<RawReading>& readings,
                                        const std::vector<double>& reference_rates) {
    if (readings.empty()) {
        return false;
    }

    // Estimate bias from zero-velocity periods or known rotations
    if (reference_rates.empty()) {
        // Zero-velocity calibration
        return calibrateGyroZeroVelocity(readings);
    } else {
        // Rotation-based calibration
        return calibrateGyroRotation(readings, reference_rates);
    }
}

bool IMUCalibration::calibrateMagnetometer(const std::vector<RawReading>& readings,
                                           const Eigen::Vector3d& reference_field) {
    if (readings.size() < 6) {
        std::cerr << "[IMUCalibration] Need at least 6 orientations for magnetometer calibration"
                  << std::endl;
        return false;
    }

    // Use ellipsoid fitting for magnetometer
    return calibrateMagEllipsoidFit(readings);
}

CalibratedReading IMUCalibration::calibrateReading(const RawReading& raw) const {
    std::lock_guard<std::mutex> lock(mutex_);

    CalibratedReading calibrated;

    if (status_ != CalibrationStatus::CALIBRATED) {
        calibrated.valid = false;
        return calibrated;
    }

    // Apply temperature compensation
    double temp_scale = 1.0;
    if (std::abs(params_.temperature_coeff) > 1e-6) {
        double temp_diff = raw.temperature - params_.reference_temperature;
        temp_scale = 1.0 + params_.temperature_coeff * temp_diff;
    }

    // Apply calibration: calibrated = scale_matrix * (raw - bias) * temp_scale
    Eigen::Vector3d corrected = raw.value - params_.bias;
    corrected *= temp_scale;
    calibrated.value = params_.scale_matrix * corrected;

    // Compute uncertainty (simplified)
    calibrated.covariance =
        Eigen::Matrix3d::Identity() * (params_.bias_uncertainty * params_.bias_uncertainty +
                                       params_.scale_uncertainty * params_.scale_uncertainty);

    calibrated.valid = true;
    return calibrated;
}

IMUCalibration::CalibrationParams IMUCalibration::getCalibrationParams() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return params_;
}

CalibrationStatus IMUCalibration::getStatus() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return status_;
}

bool IMUCalibration::isValid() const {
    return getStatus() == CalibrationStatus::CALIBRATED;
}

void IMUCalibration::reset() {
    std::lock_guard<std::mutex> lock(mutex_);
    status_ = CalibrationStatus::NOT_CALIBRATED;
    calibration_data_.clear();
    recent_readings_.clear();
    params_ = CalibrationParams{};
    params_.scale_matrix = Eigen::Matrix3d::Identity();
    params_.misalignment = Eigen::Matrix3d::Identity();
}

void IMUCalibration::estimateBias(const std::vector<RawReading>& readings) {
    if (readings.empty()) {
        return;
    }

    Eigen::Vector3d bias_sum = Eigen::Vector3d::Zero();
    for (const auto& reading : readings) {
        bias_sum += reading.value;
    }
    params_.bias = bias_sum / static_cast<double>(readings.size());
}

bool IMUCalibration::calibrateAccelStatic(const std::vector<RawReading>& readings) {
    // For accelerometer static calibration:
    // We know gravity magnitude (9.81 m/s²)
    // At each orientation, |accel| should equal gravity

    const double g = 9.81;
    std::vector<CalibrationPoint> points;

    for (const auto& reading : readings) {
        CalibrationPoint point;
        point.raw = reading;
        // Reference is gravity vector in sensor frame
        // Magnitude should be g, direction depends on orientation
        double magnitude = reading.value.norm();
        if (magnitude > 0.1) {
            point.reference = reading.value.normalized() * g;
            point.reference_uncertainty = 0.1;  // 0.1 m/s² uncertainty
            points.push_back(point);
        }
    }

    if (points.size() < 6) {
        return false;
    }

    estimateScaleAndMisalignment(points);
    return true;
}

bool IMUCalibration::calibrateGyroZeroVelocity(const std::vector<RawReading>& readings) {
    // For zero-velocity periods, gyro should read zero
    estimateBias(readings);

    // Estimate scale factors from variance
    Eigen::Vector3d variance = Eigen::Vector3d::Zero();
    for (const auto& reading : readings) {
        Eigen::Vector3d diff = reading.value - params_.bias;
        variance += diff.cwiseProduct(diff);
    }
    variance /= static_cast<double>(readings.size());

    // Scale matrix is identity with scale factors from variance
    params_.scale_matrix = Eigen::Matrix3d::Identity();
    // Scale factors inversely related to variance (higher variance = lower confidence)

    return true;
}

bool IMUCalibration::calibrateGyroRotation(const std::vector<RawReading>& readings,
                                           const std::vector<double>& reference_rates) {
    if (readings.size() != reference_rates.size()) {
        return false;
    }

    // Estimate bias and scale from known rotation rates
    estimateBias(readings);

    // Estimate scale factors
    Eigen::Vector3d scale = Eigen::Vector3d::Ones();
    for (size_t i = 0; i < readings.size(); ++i) {
        Eigen::Vector3d corrected = readings[i].value - params_.bias;
        double magnitude = corrected.norm();
        if (magnitude > 0.01 && std::abs(reference_rates[i]) > 0.01) {
            scale = scale.cwiseProduct(Eigen::Vector3d::Constant(reference_rates[i] / magnitude));
        }
    }

    params_.scale_matrix = scale.asDiagonal();
    return true;
}

bool IMUCalibration::calibrateMagEllipsoidFit(const std::vector<RawReading>& readings) {
    // Ellipsoid fitting for magnetometer calibration
    // Fit to: (x - center)^T * A * (x - center) = 1

    if (readings.size() < 6) {
        return false;
    }

    // Simple ellipsoid fit (hard iron and soft iron correction)
    // Hard iron: bias correction
    estimateBias(readings);

    // Soft iron: scale and misalignment
    // For now, use identity scale matrix
    // Full ellipsoid fitting would require solving for 9 parameters

    params_.scale_matrix = Eigen::Matrix3d::Identity();
    return true;
}

void IMUCalibration::estimateScaleAndMisalignment(const std::vector<CalibrationPoint>& points) {
    if (points.size() < 6) {
        return;
    }

    // Least squares estimation of scale matrix
    // Solve: calibrated = scale_matrix * (raw - bias)
    // Or: reference = scale_matrix * (raw - bias)

    Eigen::MatrixXd A(points.size() * 3, 9);
    Eigen::VectorXd b(points.size() * 3);

    for (size_t i = 0; i < points.size(); ++i) {
        const auto& point = points[i];
        Eigen::Vector3d corrected = point.raw.value - params_.bias;

        // Build linear system for scale matrix elements
        // For each component: ref_i = sum_j(scale_ij * corrected_j)
        for (int row = 0; row < 3; ++row) {
            int idx = i * 3 + row;
            b(idx) = point.reference(row);

            for (int col = 0; col < 3; ++col) {
                A(idx, row * 3 + col) = corrected(col);
            }
        }
    }

    // Solve least squares
    Eigen::VectorXd scale_vec = A.colPivHouseholderQr().solve(b);

    // Reshape to matrix
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            params_.scale_matrix(i, j) = scale_vec(i * 3 + j);
        }
    }
}

double IMUCalibration::computeCalibrationQuality() const {
    if (calibration_data_.empty()) {
        return 0.0;
    }

    // Compute quality based on residuals
    double total_error = 0.0;
    int count = 0;

    for (const auto& point : calibration_data_) {
        auto calibrated = calibrateReading(point.raw);
        if (calibrated.valid) {
            Eigen::Vector3d error = calibrated.value - point.reference;
            total_error += error.norm();
            count++;
        }
    }

    if (count == 0) {
        return 0.0;
    }

    double avg_error = total_error / count;
    // Quality is inverse of error (normalized)
    double quality = 1.0 / (1.0 + avg_error);
    return std::clamp(quality, 0.0, 1.0);
}

bool IMUCalibration::detectDrift(const RawReading& reading) const {
    // Simple drift detection: check if reading is far from expected
    // This is a placeholder - real implementation would use statistical tests

    if (recent_readings_.size() < 10) {
        return false;
    }

    // Compute mean and std of recent readings
    Eigen::Vector3d mean = Eigen::Vector3d::Zero();
    for (const auto& r : recent_readings_) {
        mean += r.value;
    }
    mean /= static_cast<double>(recent_readings_.size());

    Eigen::Vector3d diff = reading.value - mean;
    double distance = diff.norm();

    // Threshold based on sensor type
    double threshold = (sensor_type_ == SensorType::MAGNETOMETER) ? 0.1 : 1.0;

    return distance > threshold;
}

// IMUCalibrationSystem implementation

IMUCalibrationSystem::IMUCalibrationSystem() {
    accel_calibrator_ = std::make_shared<IMUCalibration>(IMUCalibration::SensorType::ACCELEROMETER);
    gyro_calibrator_ = std::make_shared<IMUCalibration>(IMUCalibration::SensorType::GYROSCOPE);
    mag_calibrator_ = std::make_shared<IMUCalibration>(IMUCalibration::SensorType::MAGNETOMETER);
}

IMUCalibrationSystem::~IMUCalibrationSystem() = default;

std::shared_ptr<IMUCalibration> IMUCalibrationSystem::getCalibrator(SensorType type) {
    switch (type) {
        case SensorType::ACCELEROMETER:
            return accel_calibrator_;
        case SensorType::GYROSCOPE:
            return gyro_calibrator_;
        case SensorType::MAGNETOMETER:
            return mag_calibrator_;
    }
    return nullptr;
}

bool IMUCalibrationSystem::calibrateAll(
    const std::vector<IMUCalibration::RawReading>& accel_readings,
    const std::vector<IMUCalibration::RawReading>& gyro_readings,
    const std::vector<IMUCalibration::RawReading>& mag_readings,
    const Eigen::Vector3d& reference_magnetic_field) {
    bool success = true;

    // Calibrate accelerometer
    if (!accel_readings.empty()) {
        for (const auto& reading : accel_readings) {
            IMUCalibration::CalibrationPoint point;
            point.raw = reading;
            // Reference will be computed in calibrateAccelerometer
            accel_calibrator_->addCalibrationPoint(point);
        }
        success &= accel_calibrator_->calibrate();
    }

    // Calibrate gyroscope
    if (!gyro_readings.empty()) {
        for (const auto& reading : gyro_readings) {
            IMUCalibration::CalibrationPoint point;
            point.raw = reading;
            gyro_calibrator_->addCalibrationPoint(point);
        }
        success &= gyro_calibrator_->calibrate();
    }

    // Calibrate magnetometer
    if (!mag_readings.empty()) {
        for (const auto& reading : mag_readings) {
            IMUCalibration::CalibrationPoint point;
            point.raw = reading;
            point.reference = reference_magnetic_field;
            point.reference_uncertainty = 0.001;  // 1 mT uncertainty
            mag_calibrator_->addCalibrationPoint(point);
        }
        success &= mag_calibrator_->calibrate();
    }

    return success;
}

IMUCalibrationSystem::CalibratedIMU IMUCalibrationSystem::calibrateIMU(
    const IMUCalibration::RawReading& accel, const IMUCalibration::RawReading& gyro,
    const IMUCalibration::RawReading& mag) const {
    CalibratedIMU calibrated;

    calibrated.accelerometer = accel_calibrator_->calibrateReading(accel);
    calibrated.gyroscope = gyro_calibrator_->calibrateReading(gyro);
    calibrated.magnetometer = mag_calibrator_->calibrateReading(mag);

    return calibrated;
}

IMUCalibrationSystem::CalibrationStatus IMUCalibrationSystem::getStatus() const {
    CalibrationStatus status;
    status.accel = accel_calibrator_->getStatus();
    status.gyro = gyro_calibrator_->getStatus();
    status.mag = mag_calibrator_->getStatus();
    return status;
}

}  // namespace calibration
}  // namespace fsw
